src/app/core/services/notifications/message-handler.service.ts

```diff
@@ -0,0 +1,44 @@
+import { Injectable } from '@angular/core'
+import { FirebaseX } from '@ionic-native/firebase-x/ngx'
+
+import { Assessment, AssessmentType } from '../../../shared/models/assessment'
+import { QuestionnaireService } from '../config/questionnaire.service'
+import { LogService } from '../misc/log.service'
+import { ScheduleService } from '../schedule/schedule.service'
+
+@Injectable()
+export class MessageHandlerService {
+  constructor(
+    public firebase: FirebaseX,
+    public logger: LogService,
+    public schedule: ScheduleService,
+    public questionnaire: QuestionnaireService
+  ) {
+    this.firebase
+      .onMessageReceived()
+      .subscribe(data => this.onMessageReceived(new Map(Object.entries(data))))
+  }
+
+  onMessageReceived(data) {
+    const action = data.get('action')
+    switch (action) {
+      case 'QUESTIONNAIRE_TRIGGER':
+        console.log('A questionnaire was triggered!')
+        return this.triggerQuestionnaire(JSON.parse(data.get('questionnaire')))
```

Comment:
What about the metadata?

Reply:
Ah yes I was just about to ask you about this. Do you know what the format of the metadata will be?

Reply:
I think it will be a flat json object like

```
"metadata": {
    "key1": "value1",
    "key2": "value2",
    ...
}
```

Reply:
Ok thanks I will add it.